core/dbt/contracts/graph/nodes.py

```diff
@@ -572,6 +572,48 @@ def depends_on_nodes(self):
     def depends_on_macros(self):
         return self.depends_on.macros
 
+
+# ====================================
+# CompiledNode subclasses
+# ====================================
+
+
+@dataclass
+class AnalysisNode(CompiledNode):
+    resource_type: NodeType = field(metadata={"restrict": [NodeType.Analysis]})
+
+
+@dataclass
+class HookNode(CompiledNode):
+    resource_type: NodeType = field(metadata={"restrict": [NodeType.Operation]})
+    index: Optional[int] = None
+
+
+@dataclass
+class ModelNode(CompiledNode):
+    resource_type: NodeType = field(metadata={"restrict": [NodeType.Model]})
+    access: AccessType = AccessType.Protected
+    constraints: List[ModelLevelConstraint] = field(default_factory=list)
+    version: Optional[NodeVersion] = None
+    latest_version: Optional[NodeVersion] = None
+
+    @property
+    def is_latest_version(self) -> bool:
+        return self.version is not None and self.version == self.latest_version
+
+    @property
+    def search_name(self):
+        if self.version is None:
+            return self.name
+        else:
+            return f"{self.name}.v{self.version}"
+
+    @property
+    def materialization_enforces_constraints(self):
```

Comment:
couple nits: 
- missing return type annotation `-> bool`
- simplifying to `return self.config.materialized in ["table", "incremental"]` feels more pythonic to me!